### PR TITLE
bpo-29334: Fix ssl.getpeercert for auto-handshake

### PR DESCRIPTION
**Pull request by tiran**
Drop handshake_done and peer_cert members from PySSLSocket struct. The
peer certificate can be acquired from *SSL directly.
SSL_get_peer_certificate() does not trigger any network activity.
Instead of manually tracking the handshake state, simply use
SSL_is_init_finished().

In combination these changes fix auto-handshake for non-blocking
MemoryBIO connections.

Signed-off-by: Christian Heimes <christian@python.org>

### CODE REVIEW

File: Modules/_ssl.c [hunk 1]
```diff
@@ -1506,25 +1495,30 @@ _ssl__SSLSocket_peer_certificate_impl(PySSLSocket *self, int binary_mode)
...
+    PyObject *result;
 
-    if (!self->handshake_done) {
+    if (!SSL_is_init_finished(self->ssl)) {
         PyErr_SetString(PyExc_ValueError,
                         "handshake not done yet");
         return NULL;
     }
-    if (!self->peer_cert)
+    peer_cert = SSL_get_peer_certificate(self->ssl);
```
Comment (alex) [inline Modules/_ssl.c:1506]: Previously we released the GIL around this -- this seems a bit silly to me, but I just wanted to note that this is a behavior change.

**Review by alex — APPROVED** [review]
Besides that comment, assuming tests pass, this LGTM.

**Comment by alex** [comment]
The appveyor failure appear unrelated.

**Comment by tiran** [comment]
Memo to me: add Misc/NEWS entry

**Pull request merged**